src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4199,382 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
 #endif // DEBUG
 }
 
+//****************************************************************************
+// emitRemoveJumpToNextInst:  Checks all jumps in the jump list to see if they are
+//   unconditional jumps generated at the end of BB_ALWAYS basic blocks and whether
+//   they are now a jump to the next instruction, this can happen when the following
+//   basic block emitted no instructions
+//
+// Returns:
+//   true if any jumps have been removed, otherwise returns false
+//
+// Assumptions:
+//    the jump list must be ordered by increasing igNum+insNo
+//
+bool emitter::emitRemoveJumpToNextInst()
+{
+#ifdef TARGET_XARCH
+#ifdef DEBUG
+    if (emitComp->verbose)
+    {
+        printf("*************** In emitRemoveJumpToNextInst()\n");
+    }
+#endif
+    if (emitJumpAlwaysList == nullptr)
+    {
+#ifdef DEBUG
+        if (EMITVERBOSE)
+        {
+            printf("emitRemoveJumpToNextInst emitJumpAlwaysList is empty, removed 0 jumps and 0 bytes\n");
+        }
+#endif
+        return false;
+    }
+#ifdef DEBUG
+    if (EMIT_INSTLIST_VERBOSE)
+    {
+        printf("\nInstruction group list before unconditional jump to next instruction removal:\n\n");
+        emitDispIGlist(true);
+    }
+    if (EMITVERBOSE)
+    {
+        printf("jump always list %u\n", emitJumpAlwaysList);
+        emitDispJumpList(emitJumpAlwaysList);
+    }
+#endif
+
+    int            removedCount           = 0;
+    UNATIVE_OFFSET removedSize            = 0;
+    instrDescJmp*  previousJmp            = nullptr;
+    insGroup*      previousTruncatedGroup = nullptr;
+    instrDescJmp*  lastInserted           = nullptr;
+#if DEBUG
+    UNATIVE_OFFSET previousJumpIgNo  = (UNATIVE_OFFSET)-1;
+    unsigned int   previousJumpInsNo = -1;
+#endif
+    assert(!emitJumpAlwaysListCompleted);
+    emitJumpAlwaysListCompleted = true;
+
+    while (emitJumpAlwaysList)
+    {
+        instrDescJmp* jmp = emitJumpAlwaysList;
+
+        emitJumpAlwaysList = jmp->idjNext;
+        jmp->idjNext       = nullptr;
+
+        // if the jump is unconditional then it is a candidate
+        if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && !jmp->idjKeepLong)
+        {
+#if DEBUG
+            assert(jmp->idjIG->igNum > previousJumpIgNo || previousJumpIgNo == (UNATIVE_OFFSET)-1 ||
+                   (jmp->idjIG->igNum == previousJumpIgNo && jmp->idDebugOnlyInfo()->idNum > previousJumpInsNo));
+            previousJumpIgNo  = jmp->idjIG->igNum;
+            previousJumpInsNo = jmp->idDebugOnlyInfo()->idNum;
+#endif
+
+            // target group is not bound yet so use the cookie to fetch it
+            insGroup* targetGroup = (insGroup*)emitCodeGetCookie(jmp->idAddr()->iiaBBlabel);
+
+            if (targetGroup != nullptr && (jmp->idjIG->igFlags & IGF_HAS_ALIGN) == 0 &&
+                jmp->idjIG->igNext == targetGroup)
+            {
+                insGroup* group            = jmp->idjIG;
```

Comment:
Cache `insGroup* group            = jmp->idjIG;` at the begining of the `emitJumpAlwaysList` loop and use it throughout the loop body?

Reply:
Also rename `group` to something meaningful like `containingIG`.

Reply:
renamed to jmpGroup, moved to the top and used throughout the loop.